dali/pipeline/operators/geometric/flip.cu

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime_api.h>
+#include <dali/pipeline/operators/geometric/flip.h>
+
+namespace dali {
+
+template <>
+Flip<GPUBackend>::Flip(const OpSpec &spec)
+    : Operator<GPUBackend>(spec),
+      _horizontal(spec.GetArgument<int32>("horizontal")),
+      _vertical(spec.GetArgument<int32>("vertical")) {}
+
+struct HorizontalIdxFlip {
+  static __device__ size_t OutputIdx(size_t height, size_t width, size_t r, size_t c) {
+    return r * width + (width - c - 1);
+  }
+};
+
+struct VerticalIdxFlip {
+  static __device__ size_t OutputIdx(size_t height, size_t width, size_t r, size_t c) {
+    return (height - r - 1) * width + c;
+  }
+};
+
+struct HorizontalVerticalIdxFlip {
+  static __device__ size_t OutputIdx(size_t height, size_t width, size_t r, size_t c) {
+    return (height - r - 1) * width + (width - c - 1);
+  }
+};
+
+template <typename IndexFlip, typename T>
+__global__ void FlipKernel(T *output, const T *input, size_t height, size_t width,
+                           size_t channels) {
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx < height * width) {
+    size_t r = idx / width;
+    size_t c = idx % width;
+    size_t out_idx = IndexFlip::OutputIdx(height, width, r, c);
+    memcpy(&output[out_idx * channels], &input[(r * width + c) * channels], sizeof(T) * channels);
+  }
+}
+
+template <typename IndexFlip>
+void RunKernel(TensorList<GPUBackend> &output, const TensorList<GPUBackend> &input,
+               cudaStream_t stream) {
+  DALI_TYPE_SWITCH(input.type().id(), DType,
+    for (unsigned int i = 0; i < input.ntensor(); ++i) {
```

Comment:
Do we want to be able to process more than 1 image per cuda kernel invokation?

Reply:
Batch processing would be nice, but coming up with a sensible block layout is not that easy. Block-per-image approach is not efficient, since we can't distribute work across more SMs than there are images in a batch.